go/store/nbs/byte_sink.go

```diff
@@ -0,0 +1,289 @@
+
+	if remaining >= srcLen {
+		currBlock = append(currBlock, src...)
+		sink.blocks[currBlockIdx] = currBlock
+	} else {
+		if remaining > 0 {
+			currBlock = append(currBlock, src[:remaining]...)
+			sink.blocks[currBlockIdx] = currBlock
+		}
+
+		newBlock := make([]byte, 0, sink.blockSize)
+		newBlock = append(newBlock, src[remaining:]...)
+		sink.blocks = append(sink.blocks, newBlock)
+	}
+
+	sink.pos += uint64(srcLen)
+	return srcLen, nil
+}
+
+// Flush writes all the data that was written to the ByteSink to the supplied writer
+func (sink *BlockBufferByteSink) Flush(wr io.Writer) (err error) {
+	return iohelp.WriteAll(wr, sink.blocks...)
+}
+
+// FlushToFile writes all teh data that was written to the file to a file at the provided location.
+func (sink *BlockBufferByteSink) FlushToFile(path string) (err error) {
+	return flushSinkToFile(sink, path)
+}
+
+// BufferedFileByteSink is a ByteSink implementation that buffers some amount of data before it passes it
+// to a background writing thread to be flushed to a file.
+type BufferedFileByteSink struct {
+	blockSize    int
+	pos          uint64
+	currentBlock []byte
+
+	writeCh chan []byte
+	ae      *atomicerr.AtomicError
+	wg      *sync.WaitGroup
+
+	wr   io.WriteCloser
+	path string
+}
+
+// NewBufferedFileByteSink creates a BufferedFileByteSink
+func NewBufferedFileByteSink(blockSize, chBufferSize int) (*BufferedFileByteSink, error) {
+	path := filepath.Join(os.TempDir(), uuid.New().String()+".tf")
+	f, err := os.OpenFile(path, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, os.ModePerm)
+
+	if err != nil {
+		return nil, err
+	}
+
+	sink := &BufferedFileByteSink{
+		blockSize:    blockSize,
+		currentBlock: make([]byte, blockSize),
+		writeCh:      make(chan []byte, chBufferSize),
+		ae:           atomicerr.New(),
+		wg:           &sync.WaitGroup{},
+		wr:           f,
+		path:         path,
+	}
+
+	sink.wg.Add(1)
+	go func() {
+		defer sink.wg.Done()
+		sink.backgroundWrite()
+	}()
+
+	return sink, nil
+}
+
+// Write writes a byte array to the sink.
+func (sink *BufferedFileByteSink) Write(src []byte) (int, error) {
+	srcLen := len(src)
+	remaining := cap(sink.currentBlock) - len(sink.currentBlock)
+
+	if remaining >= srcLen {
+		sink.currentBlock = append(sink.currentBlock, src...)
+
+		if remaining == srcLen {
+			sink.writeCh <- sink.currentBlock
+			sink.currentBlock = nil
+		}
+	} else {
+		if remaining > 0 {
+			sink.currentBlock = append(sink.currentBlock, src[:remaining]...)
+			sink.writeCh <- sink.currentBlock
+		}
+
+		newBlock := make([]byte, 0, sink.blockSize)
+		newBlock = append(newBlock, src[remaining:]...)
+		sink.currentBlock = newBlock
+	}
+
+	sink.pos += uint64(srcLen)
+	return srcLen, nil
+}
+
+func (sink *BufferedFileByteSink) backgroundWrite() {
+	var err error
+	for buff := range sink.writeCh {
+		if err != nil {
+			continue // drain
+		}
+
+		err = iohelp.WriteAll(sink.wr, buff)
+		sink.ae.SetIfError(err)
+	}
+
+	err = sink.wr.Close()
```

Comment:
This stomps any previously set error

Reply:
that's fine. It's just updating the ae which won't change if it's already set.